src/data/variationOfCycleChecker.ts

```diff
@@ -0,0 +1,70 @@
+import {DbTricksTableZod} from "../lib/database/schemas/CurrentVersionSchema"
+import { z } from "zod";
+
+type Trick = z.infer<typeof DbTricksTableZod>
+
+/**
+ * Tricks are set up in a way where variationOf defines a "parent" trick they belong to. 
+ * Important note here is that we do not have trees as a trick can have multiple parents.
+ * 
+ * As such, the set of tricks can be described as a set of disjunct directed cycle-free graphs.
+ * This module is responsible for detecting cycles.
+ *  
+ */
+export default function buildGraph(tricks: Trick[]) {
+
+  /**
+   * This is a "shortcut" map telling us all the roots of a trick. Mainly
+   * here to skip doing redundant work. If a path has already been taken, it can be skipped
+   */
+  const rootsIndex: Record<Trick["id"], Trick["id"][]> = {}
+
+  /**
+   * Create a map of child ID -> parent ID for easy lookup
+   */
+  const childIdToParentIdsLookup: Record<Trick["id"], Trick["id"][]> = {}
+
+  tricks.forEach( e => {
+    const parents = (e.variationOf ?? []).map( e => e[0])
+    childIdToParentIdsLookup[e.id] = parents
+  })
+
+  // A recursive function that looks at the immediate parent nodes 
+  function handleTraversal(currentId: number, currentChain: Trick["id"][]) {
```

Comment:
This function could be extracted from `buildGraph` to keep it short and concise

Reply:
I would keep the function as is — it makes use of data outside the `handleTraversal` scope.
![image](https://github.com/bastislack/highline-freestyle/assets/27686785/287bf093-d269-4951-aa53-82c0e16780e0)